### ISSUE TITLE
Dynamo dynamic shapes reduction guard failure

### ISSUE THREAD

**Issue by anwang2009** [pytorch/pytorch#95981]
### 🐛 Describe the bug

With dynamic shapes enabled, I'm seeing an unexpected guard failure causing a recompilation on a simple softmax:
```
import torch
from torch import _dynamo as dynamo

def f(a):
    return torch.softmax(a, dim=0)

def custom(gm, example_inputs):
    print("Recompiling")
    return gm.forward

def guard_fail(failure):
    print("Guard failure", failure)

f = dynamo.optimize(custom, dynamic=True, guard_fail_fn=guard_fail)(f)
f(torch.Tensor(2))
f(torch.Tensor(3))
```

It looks like the culprit is here: https://github.com/pytorch/pytorch/blob/4026c62174a4670c95148f6ff9318637d6f38c60/torch/_refs/__init__.py#L2070

Changing 

```valid_shape = a.ndim == 0 or py_all(a.shape[i] for i in dims)```

to

```valid_shape = a.ndim == 0 or py_all(a.shape[i] != 0 for i in dims)```

appears to resolve the bug.

### Versions

```
[pip3] numpy==1.24.2
[pip3] pytorch-triton==2.0.0+b8b470bc59
[pip3] torch==2.1.0.dev20230303+cu117
[pip3] torchaudio==2.0.0.dev20230303+cu117
[pip3] torchvision==0.15.0.dev20230303+cu117
[conda] numpy                     1.24.2                   pypi_0    pypi
[conda] pytorch-triton            2.0.0+b8b470bc59          pypi_0    pypi
[conda] torch                     2.1.0.dev20230303+cu117          pypi_0    pypi
[conda] torchaudio                2.0.0.dev20230303+cu117          pypi_0    pypi
[conda] torchvision               0.15.0.dev20230303+cu117          pypi_0    pypi
```

cc @ezyang @soumith @msaroufim @wconstab @ngimel @bdhirsh

**Comment by anwang2009** [comment]
The underlying issue is the current behavior of `SymNode._bool` -- in this case it calls `guard_bool("", 0)` which forces an evaluation of `node.expr` and inserts a guard. Is the desired approach to fix every operator that evaluates bools on shapes like the suggested resolution above? Or is there something we can do with `SymNode._bool` itself?

**Comment by ngimel** [comment]
More general solution would be to make SymNode._bool not create overly restrictive guards, but we'd accept solution above to unblock, since it's not exactly clear how to do a more general thing.

**Comment by ezyang** [comment]
The general solution should be pretty simple. When we convert a SymInt to bool (as is done here), we should do a `x == 0` test (not a guard on the exact value of the int instead.)
